pkg/reconciler/pipelinerun/affinity_assistant_test.go

```diff
@@ -82,6 +82,55 @@ func TestCreateAndDeleteOfAffinityAssistant(t *testing.T) {
 	}
 }
 
+func TestThatTolerationsAndNodeSelectorIsPropagatedToAffinityAssistant(t *testing.T) {
+	prWithCustomPodTemplate := &v1beta1.PipelineRun{
+		TypeMeta: metav1.TypeMeta{Kind: "PipelineRun"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "pipelinerun-with-custom-podtemplate",
+		},
+		Spec: v1beta1.PipelineRunSpec{
+			PodTemplate: &v1beta1.PodTemplate{
+				NodeSelector: map[string]string{
+					"disktype": "ssd",
+				},
+				Tolerations: []corev1.Toleration{{
+					Key:      "key",
+					Operator: "Equal",
+					Value:    "value",
+					Effect:   "NoSchedule",
+				}},
+			},
+		},
+	}
+
+	prWithoutCustomPodTemplate := &v1beta1.PipelineRun{
+		TypeMeta: metav1.TypeMeta{Kind: "PipelineRun"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "pipelinerun-without-custom-podtemplate",
+		},
+		Spec: v1beta1.PipelineRunSpec{},
+	}
+
+	stsWithTolerationsAndNodeSelector := affinityAssistantStatefulSet("test-assistant", prWithCustomPodTemplate, "mypvc")
+	stsWithoutTolerationsAndNodeSelector := affinityAssistantStatefulSet("test-assistant", prWithoutCustomPodTemplate, "mypvc")
+
+	if len(stsWithTolerationsAndNodeSelector.Spec.Template.Spec.NodeSelector) != 1 {
+		t.Errorf("expected a NodeSelector in the StatefulSet")
+	}
+
+	if len(stsWithTolerationsAndNodeSelector.Spec.Template.Spec.Tolerations) != 1 {
+		t.Errorf("expected Tolerations in the StatefulSet")
+	}
+
+	if len(stsWithoutTolerationsAndNodeSelector.Spec.Template.Spec.NodeSelector) != 0 {
+		t.Errorf("unexpected NodeSelector in the StatefulSet")
+	}
+
+	if len(stsWithoutTolerationsAndNodeSelector.Spec.Template.Spec.Tolerations) != 0 {
+		t.Errorf("unexpected Tolerations in the StatefulSet")
```

Comment:
it feels like this test include 2 test cases in one - is it possible to create two different Test functions, one for each case? (one for with tolerations, one for without)

Reply:
Sure, I can do that.

Reply:
my apologies but i have 1 more request around the tests! now that I see them separated, it feels to me like the original test was actually 4 different cases:

- node selector in pod template
- node selector not in pod template
- tolerations in pod template
- tolerations not in pod template

now that we have 2 test functions, we've had to duplicate the test data (with and without)

i suggest that we either:
1. have 1 test for prWithCustomPodTemplate only, and one test for prWithoutCustomPodTemplate. The with case can check both node selector and tolerations have progated, the without can check they havent (this will allow us to not need to duplicate the test input)
2. keep the separate tests the way they are but use table driven tests such that each of prWithCustomPodTemplate and prWithoutCustomPodTemplate are a separate test case (i.e. we end up with 4 tests total, 2 test functions) 

my preference is for (1)

apologies for nitpicking this test

Reply:
I agree. By not duplicating test data, it makes it more readable. Thanks, good feedback make me a better developer. The test is now updated.